### PR TITLE
feat(sla): add month duration

### PR DESCRIPTION
**Pull request by stonebuzz**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

File: tests/functionnal/SLM.php [hunk 1]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+      $slm_id = $slm->add(
+         [
+            'name'         => 'Test SLM',
+            'calendars_id' => 0, //24/24 7/7
+         ]
+      );
+      $this->integer($slm_id)->isGreaterThan(0);
+
+      $ola_tto = new \OLA();
+      $ola_ttr_id = $ola_tto->add(
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:643]:
```suggestion
      $ola_ttr = new \OLA();
      $ola_ttr_id = $ola_ttr->add(
```

File: tests/functionnal/SLM.php [hunk 2]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+            'type'               => \SLM::TTR,
+            'number_time'        => 1,
+            'definition_time'    => 'month',
+            'end_of_working_day' => false,
+         ]
+      );
+      $this->integer($ola_ttr_id)->isGreaterThan(0);
+
+      $ola_ttr = new \OLA();
+      $ola_tto_id = $ola_ttr->add(
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:656]:
```suggestion
      $ola_tto = new \OLA();
      $ola_tto_id = $ola_tto->add(
```

File: tests/functionnal/SLM.php [hunk 3]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+            'type'               => \SLM::TTO,
+            'number_time'        => 1,
+            'definition_time'    => 'month',
+            'end_of_working_day' => false,
+         ]
+      );
+      $this->integer($ola_tto_id)->isGreaterThan(0);
+
+      $sla_tto = new \SLA();
+      $sla_ttr_id = $sla_tto->add(
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:669]:
```suggestion
      $sla_ttr = new \SLA();
      $sla_ttr_id = $sla_ttr->add(
```

File: tests/functionnal/SLM.php [hunk 4]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+            'type'               => \SLM::TTR,
+            'number_time'        => 1,
+            'definition_time'    => 'month',
+            'end_of_working_day' => false,
+         ]
+      );
+      $this->integer($sla_ttr_id)->isGreaterThan(0);
+
+      $sla_ttr = new \SLA();
+      $sla_tto_id = $sla_ttr->add(
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:682]:
```suggestion
      $sla_tto = new \SLA();
      $sla_tto_id = $sla_tto->add(
```

File: tests/functionnal/SLM.php [hunk 5]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+      $this->variable($ticket->fields['internal_time_to_resolve'])->isEqualTo($after1month);
+
+      $this->integer((int)$ticket->fields['olas_id_tto'])->isEqualTo($ola_tto_id);
+      $this->variable($ticket->fields['internal_time_to_own'])->isEqualTo($after1month);
+
+      $this->integer((int)$ticket->fields['slas_id_ttr'])->isEqualTo($sla_ttr_id);
+      $this->variable($ticket->fields['time_to_resolve'])->isEqualTo($after1month);
+
+      $this->integer((int)$ticket->fields['slas_id_tto'])->isEqualTo($sla_tto_id);
+      $this->variable($ticket->fields['time_to_own'])->isEqualTo($after1month);
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:723]: I would test different dates for each field. In deed, if there is a mixup between OLA/SLA TTR/TTO, this test will not detect it.

File: tests/functionnal/SLM.php [hunk 6]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+
+      $this->integer(strtotime($ticket->fields['ola_ttr_begin_date']))
+         ->isGreaterThanOrEqualTo($createtime)
+         ->isLessThanOrEqualTo($createtime);
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:727]: `>= $createtime && <= $createtime` -> isEqualTo ?

May be you can check that it is equal to `date` field of the ticket.

File: tests/functionnal/SLM.php [hunk 7]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+      );
+      $this->integer($ticket_id)->isGreaterThan(0);
+
+      $this->boolean($ticket->getFromDB($ticket_id))->isTrue();
+      $this->integer((int)$ticket->fields['olas_id_ttr'])->isEqualTo(0);
+      $this->variable($ticket->fields['ola_ttr_begin_date'])->isEqualTo(null);
+      $this->variable($ticket->fields['internal_time_to_resolve'])->isEqualTo(null);
+
+      //Wait...
+      sleep(1);
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:745]: I think you can force opening date of the ticket by setting `date` field on `$ticket->add()` input.

File: tests/functionnal/SLM.php [hunk 8]
```diff
@@ -399,4 +622,158 @@ public function testInternalTtrComputation() {
...
+
+      $this->integer((int)$ticket->fields['slas_id_ttr'])->isEqualTo($sla_ttr_id);
+      $this->variable(date('Y-m-d H:i', strtotime($ticket->fields['time_to_resolve'])))->isEqualTo($after1month);
+
+      $this->integer((int)$ticket->fields['slas_id_tto'])->isEqualTo($sla_tto_id);
+      $this->variable(date('Y-m-d H:i', strtotime($ticket->fields['time_to_own'])))->isEqualTo($after1month);
+
+      $this->integer(strtotime($ticket->fields['ola_ttr_begin_date']))
+         ->isGreaterThanOrEqualTo($update_time_1)
+         ->isLessThanOrEqualTo($update_time_2);
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:776]: May be you can check that it is equal to `date_mod` field of the ticket.

File: tests/functionnal/SLM.php [hunk 9]
```diff
@@ -270,6 +270,229 @@ public function testLifecyle() {
...
+      $this->checkInput($cal, $cal_id);
+      for ($day = 1; $day <= 5; $day++) {
+         $calseg_id = $calseg->add([
+            'calendars_id' => $cal_id,
+            'day'          => $day,
+            'begin'        => '08:00:00',
+            'end'          => '20:00:00'
+         ]);
+         $this->checkInput($calseg, $calseg_id);
+      }
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:296]: You can use the default GLPI calendar:
```suggestion
      $cal_id = getItemByTypeName('Calendar', 'Default', true);
```

File: tests/functionnal/SLM.php [hunk 10]
```diff
@@ -270,6 +270,229 @@ public function testLifecyle() {
...
+      $this->integer((int)$ticket->getField('olas_id_ttr'))->isEqualTo($ola2_id);
+      $this->string($ticket->getField('time_to_resolve'))->length->isEqualTo(19);
+
+      // test update ticket
+      $ticket = new \Ticket;
+      $tickets_id_2 = $ticket->add($ticket_input_2 = [
+         'name'    => "to be updated",
+         'content' => $this->method
+      ]);
+      $ticket->update([
```
Comment (cedric-anne) [inline tests/functionnal/SLM.php:462]: You should check that OLA/SLA fields are not already set before update.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: tests/functionnal/SLM.php (hunk 1 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 2 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 3 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 4 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 6 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 7 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 8 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 9 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

File: tests/functionnal/SLM.php (hunk 10 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php:458]: Fixed

File: tests/functionnal/SLM.php (hunk 5 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: Fixed

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**